src/ray/object_manager/object_manager.cc

```diff
@@ -1,156 +1,309 @@
+      new ObjectStoreNotification(main_service, config.store_socket_name));
+  store_notification_->SubscribeObjAdded(
       [this](const ObjectID &oid) { NotifyDirectoryObjectAdd(oid); });
-  store_client_->SubscribeObjDeleted(
+  store_notification_->SubscribeObjDeleted(
       [this](const ObjectID &oid) { NotifyDirectoryObjectDeleted(oid); });
   StartIOService();
 };
 
 void ObjectManager::StartIOService() {
   io_thread_ = std::thread(&ObjectManager::IOServiceLoop, this);
-  //  thread_group_.create_thread(boost::bind(&boost::asio::io_service::run,
-  //  &io_service_));
+  // thread_group_.create_thread(boost::bind(&asio::io_service::run,
+  // &io_service_));
 }
 
-void ObjectManager::IOServiceLoop() { io_service_.run(); }
+void ObjectManager::IOServiceLoop() { object_manager_service_->run(); }
 
 void ObjectManager::StopIOService() {
-  io_service_.stop();
+  object_manager_service_->stop();
   io_thread_.join();
-  //  thread_group_.join_all();
+  // thread_group_.join_all();
 }
 
-void ObjectManager::SetClientID(const ClientID &client_id) { client_id_ = client_id; }
-
-ClientID ObjectManager::GetClientID() { return client_id_; }
-
 void ObjectManager::NotifyDirectoryObjectAdd(const ObjectID &object_id) {
+  local_objects_.insert(object_id);
   ray::Status status = object_directory_->ReportObjectAdded(object_id, client_id_);
 }
 
 void ObjectManager::NotifyDirectoryObjectDeleted(const ObjectID &object_id) {
+  local_objects_.erase(object_id);
   ray::Status status = object_directory_->ReportObjectRemoved(object_id, client_id_);
 }
 
 ray::Status ObjectManager::Terminate() {
   StopIOService();
   ray::Status status_code = object_directory_->Terminate();
   // TODO: evaluate store client termination status.
-  store_client_->Terminate();
+  store_notification_->Terminate();
+  store_pool_->Terminate();
   return status_code;
 };
 
 ray::Status ObjectManager::SubscribeObjAdded(
     std::function<void(const ObjectID &)> callback) {
-  store_client_->SubscribeObjAdded(callback);
+  store_notification_->SubscribeObjAdded(callback);
   return ray::Status::OK();
 };
 
 ray::Status ObjectManager::SubscribeObjDeleted(
     std::function<void(const ObjectID &)> callback) {
-  store_client_->SubscribeObjDeleted(callback);
+  store_notification_->SubscribeObjDeleted(callback);
   return ray::Status::OK();
 };
 
 ray::Status ObjectManager::Pull(const ObjectID &object_id) {
-  // TODO(hme): Need to correct. Workaround to get all pull requests on the same thread.
-  SchedulePull(object_id, 0);
+  main_service_->post([this, object_id]() { RAY_CHECK_OK(Pull_(object_id)); });
   return Status::OK();
 };
 
 void ObjectManager::SchedulePull(const ObjectID &object_id, int wait_ms) {
-  pull_requests_[object_id] = Timer(new boost::asio::deadline_timer(
-      io_service_, boost::posix_time::milliseconds(wait_ms)));
+  pull_requests_[object_id] = Timer(
+      new asio::deadline_timer(*main_service_, boost::posix_time::milliseconds(wait_ms)));
```

Comment:
The `typedef` is removed, but rewriting the dynamic allocation of timers will take time and I would like to understand the justification for this before refactoring.

Reply:
As a general rule of thumb we shouldn't be using `new`/`delete` unless there is a very good reason.

That way, the code will be a lot cleaner and less error prone since resource management will be automatic. There are also other benefits.

https://stackoverflow.com/questions/6500313/why-should-c-programmers-minimize-use-of-new
https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization
https://google.github.io/styleguide/cppguide.html#Ownership_and_Smart_Pointers

Reply:
Is this the only reason? `Timer` is a typedef of a shared pointer (I've since removed the typedef).